### PR TITLE
Replace core reflection usage with MethodHandles

### PR DESCRIPTION
**Pull request by stevenschlansker**
This should improve performance due to the new design of method handles.
Core reflection has to evaluate everything at the time that you call `method.invoke` - it has to do access checks, determine what parameter conversions are necessary, etc - and does not have any local state to save the result of that work.

With MethodHandle, you do more of the work up-front - determine access, determine what parameter boxing or casts are necessary - and store that computed state in a new MethodHandle.  Later, when you `invokeExact`, the method handle is able to make the call much faster.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -142,6 +142,11 @@
       <artifactId>junit-jupiter-api</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (stevenschlansker) [inline pom.xml:145]: I got some errors about `junit-jupiter-params` being used but not declared e.g. `EnumNamingStrategiesTest` - do we need a dependency here?

File: src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java [hunk 2]
```diff
@@ -66,24 +66,15 @@ Creators collect(Class<?> primaryMixIn)
...
-            for (int i = factories.size(); --i >= 0; ) {
-                if (_intr.hasIgnoreMarker(_config, factories.get(i))) {
-                    factories.remove(i);
-                }
-            }
+        if (_collectAnnotations && _defaultConstructor != null && _intr.hasIgnoreMarker(_config, _defaultConstructor)) {
+            _defaultConstructor = null;
         }
+
+        constructors.removeIf(constructor ->
```
Comment (stevenschlansker) [inline src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java:73]: `removeIf` should perform better than `remove` in a loop - the `ArrayList` implementation collects a bitmask of elements to remove and then does it in a single copy, rather than repeated shifting

File: src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java [hunk 3]
```diff
@@ -22,6 +23,7 @@ public UnmodifiableCollectionMixin(final Collection<?> collection) { }
     }
 
     @Test
+    @Disabled // XXX(scs) What to do about this test?
```
Comment (stevenschlansker) [inline src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: This test is kinda crazy - it deserializes an `java.util.Collections$UnmodifiableCollection` which is not a public type, and the arguments don't even match the constructor signature... I'm not sure how it is supposed to work

File: src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java (hunk 2 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java:73]: Sounds good -- I can merge this separately?

File: pom.xml (hunk 1 above)
Comment (cowtowncoder) [reply pom.xml:145]: Let me check and possible merge separately

Comment (cowtowncoder) [reply pom.xml:145]: Merged. Guessing it came as transitive dependency, but module-info.java declared requirement anyway.

File: src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java (hunk 3 above)
Comment (cowtowncoder) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: Ability to access JDK internal types worked for JDK 8, but things are changing.

But as to this:

> arguments don't even match the constructor signature

I don't remember details but this test has been passing on JDKs up to 21.

Mix-in matching only consider number of parameters and type-erased types of parameters (trying to match what JVM/JDK do). How does actual Constructor differ when you see it?

**Comment by cowtowncoder** [comment]
Also looks like Android compatibility is broken? (wrt animal-sniffer failure for JDK 17 build)

**Comment by pjfanning** [comment]
Java 17 compile issues
```
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/ser/BeanPropertyWriter.java:614: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/ser/BeanPropertyWriter.java:6[9](https://github.com/FasterXML/jackson-databind/actions/runs/14025726660/job/39263945347?pr=5046#step:6:10)3: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/ser/BeanPropertyWriter.java:814: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/util/ClassUtil.java:859: Undefined reference: boolean java.lang.reflect.AccessibleObject.trySetAccessible()
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:161: Undefined reference: void java.lang.invoke.MethodHandle.invokeExact(Object, Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:190: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object, Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:208: Undefined reference: void java.lang.invoke.MethodHandle.invokeExact(Object, Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:225: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object, Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:[12](https://github.com/FasterXML/jackson-databind/actions/runs/14025726660/job/39263945347?pr=5046#step:6:13)7: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact()
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:148: Undefined reference: Object java.lang.invoke.MethodHandle.invokeExact(Object)
Error:  /home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:160: Undefined reference: Object java.lang.invoke.MethodHandle.invoke(Object)
Error:  Failed to execute goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.24:check (default-cli) on project jackson-databind: Signature errors found. Verify them and ignore them with the proper annotation if needed. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Error: Process completed with exit code 1.
```

**Comment by pjfanning** [comment]
> > Java 17 compile issues
> 
> I believe those animal sniffer warnings are erroneous & can be suppressed: [mojohaus/animal-sniffer#67](https://github.com/mojohaus/animal-sniffer/issues/67)
> 
> also the min android sdk for MethodHandles is 26, which is already applied

Can't we increase the min Android SDK?

**Comment by iProdigy** [comment]
> Can't we increase the min Android SDK?

To clarify, jackson 3 declares compatibility with sdk 34+ while 2.18 declares sdk 26+, both of which support MethodHandles - I don't follow your question

**Comment by cowtowncoder** [comment]
Maybe there are newer versions of Animal sniffer configs (dep version from pom.xml)?

**Comment by yawkat** [comment]
A new sniffer config is not sufficient, as PolymorphicSignature methods can have an unlimited number of signatures. It needs real plugin support.

I also don't think we can expect a response from the maintainers anytime soon, judging by that issue and the linked issues. Netty even removed the sniffer plugin because of it: https://github.com/netty/netty/pull/14032

I think the the only sensible path forward is to add an exclusion for MethodHandle for now

**Comment by cowtowncoder** [comment]
@yawkat Ok. I will update dependency just as routine maintenance (for 2.19 and master, both behind a bit), but not expecting this particular issue to be addressed.
It does sound like an override needed then.

**Comment by stevenschlansker** [comment]
I added an ignore for MethodHandle.
Now, the remaining AnimalSniffer problem is:
```
/home/runner/work/jackson-databind/jackson-databind/src/main/java/tools/jackson/databind/util/ClassUtil.java:859: Undefined reference: boolean java.lang.reflect.AccessibleObject.trySetAccessible()
```

I checked the definition and it seems that the method is part of the java spec since Java 9. Surely Android is past Java 9 baseline at this point - should I add another ignore for this one too?

File: pom.xml (hunk 1 above)
Comment (stevenschlansker) [reply pom.xml]: Thanks!

File: src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java (hunk 2 above)
Comment (stevenschlansker) [reply src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java:73]: Sounds good. There is a slight behavior change here too - since some methods are not accessible, even if we try to `setAccessible`, the `null` filter is now unconditional whether we are looking for `@JsonIgnore` or not.

**Comment by iProdigy** [comment]
> Surely Android is past Java 9 baseline at this point - should I add another ignore for this one too?

android doesn't seem to implement that method on any SDK: https://developer.android.com/reference/java/lang/reflect/AccessibleObject

you'll have to switch to `setAccessible` (and catch the exception)

**Comment by stevenschlansker** [comment]
Rats, that sucks. I'll make that update...

**Comment by stevenschlansker** [comment]
Ok, I reworked the `ClassUtil.checkAndFixAccess` method. There is a behavior change - previously, if we could not `setAccessible` on a member, we would end up throwing an exception. Now, we just skip the property, since I kept running into things like random attempts to crack open e.g. `sun.util.calendar.ZoneInfo` which fails. LMK if this is a problem and we can discuss more.

**Comment by stevenschlansker** [comment]
WTF, Android does not implement `InaccessibleObjectException` either?!

File: src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java (hunk 3 above)
Comment (stevenschlansker) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: Well what surprised me I guess is that the constructor itself takes a `Collection` -- but I guess Jackson is smart enough to first allocate e.g. `ArrayList` for the elements, then forcibly create an `UnmodifiableCollection`.

I think in order to play by the rules, we'd need to somehow map `java.util.Collections$UnmodifiableCollection` and all its friends to their static factory methods like `Collections.unmodifiableCollection` rather than forcing open the constructor. I'm not sure how to do that other than to create a special rule that maps all the `Collections` factory methods. Probably same for `synchronized`. Do you have another idea or better suggestion?

Comment (stevenschlansker) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: It's also weird because we end up writing the non-public class name `java.util.Collections$UnmodifiableCollection` to the wire, which is not a stable public name, and in theory could change release to release.

Do users really expect to serialize collections and re-construct the unmodifiable / synchronized wrappers? Seems like a very fragile expectation to me...

**Comment by cowtowncoder** [comment]
> Ok, I reworked the `ClassUtil.checkAndFixAccess` method. There is a behavior change - previously, if we could not `setAccessible` on a member, we would end up throwing an exception. Now, we just skip the property, since I kept running into things like random attempts to crack open e.g. `sun.util.calendar.ZoneInfo` which fails. LMK if this is a problem and we can discuss more.

Hmm. Ok, that might be ok. Was thinking that maybe "sun.*" was not checked as system class by `ClassUtil.isJDKClass()` (and if so, could just add), but it is included.

I think it'd be good to try to avoid these calls to be sure (instead of having to catch exception) but that might be yet bigger undertaking.

File: src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java (hunk 3 above)
Comment (cowtowncoder) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: As the background, I _think_ this was to support use case that happened to be working, so gnarly implementation / test case was added to guard against regression.
I am not sure it is reasonable to expect class-name-based polymorphic handling to work seamlessly on modern JDKs -- I don't think we should have super-elaborate implementation.
(plus here we are relying on specific class name for case where I'm sure JDK is not required to keep it. But relying on class names for serialization is fragile no matter what I guess).

Comment (cowtowncoder) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: But, beyond this, whether users expect this -- I don't know. Some probably do. Do we need to try really hard to maintain such invariants. Probably not.

Comment (stevenschlansker) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: Ok, thanks for considering this. Personally I would vote that this is too weird of a use case to support - the mapping from the non-public collection class to its public static factory is not mechanical, we would need to maintain a mapping by hand or clever algorithm, and Jackson 3 is a good time to consider dropping some legacy features that might be ill-advised in the new world of "secure by default" Java.

JSON is not really intended to be Java serialization - which preserves the entire object graph - JSON really is "here are the elements of the collection", not, "use this specific private implementation of Java collections".

If a user comes back and complains that something reasonable does not work, we could pick it back up then. Does that sound good to you? Or would you have a different preferred outcome here?

**Comment by stevenschlansker** [comment]
> I think it'd be good to try to avoid these calls to be sure (instead of having to catch exception) but that might be yet bigger undertaking.

Yeah, that's why I tried to replace it with `trySetAccessible` - while it is still overriding visibility rules, at least it fails normally instead of requiring an exception thrown / catch.

The good news is this only happens during inspection, not at deserialization time, so the performance impact would be muted.

File: src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java (hunk 3 above)
Comment (cowtowncoder) [reply src/test/java/tools/jackson/databind/mixins/MixinForCreators2795Test.java:26]: I agree with. While it would be nice in some ways to support all existing use cases, there must be some limit. And this, I think, falls beyond that limit. So let's leave it `@Disabled` and go with that.

**Comment by cowtowncoder** [comment]
@stevenschlansker Excellent work! I am bit hesitant to merge this before 3.0.0-rc2, but could be merged right after -- esp. assuming we could get some performance numbers, maybe for https://github.com/FasterXML/jackson-benchmarks/ (I'll see if I could run some myself).

**Comment by yawkat** [comment]
Maybe you can use a method handle to call trySetInvisible :smile:

**Comment by stevenschlansker** [comment]
Thanks @cowtowncoder I am excited to finally get this work in :)
I rebased and squished the MR. Please let me know if there's any other tasks necessary pre-merge.

> I am bit hesitant to merge this before 3.0.0-rc2, but could be merged right after

That's fine by me. Were I making the call, I'd think that having the change in users' hands in an earlier RC for more testing would be better than delaying merging and finding problems closer to release - but again, 100% your call :)

> EDIT: from quick test runs, it looks like there's maybe +5% for regular JSON read via databind for this PR vs 3.0.0-rc1.
> Hoping to test out writes too.
> So maybe somewhat limited benefit at least for tested case.

I'm not too surprised. There was a time, a long time ago, when reflection was quite slow. The JVM has come a long way and now reflective calls are much, much faster than they used to be. That said, 5% is 5%, so I will take it :) especially since it comes with little downside.

> Maybe you can use a method handle to call trySetAccessible

@yawkat yes, that would work :) I think since this is called only during (de)serializer instantiation, though, it's not worth the effort for now.

**Comment by cowtowncoder** [comment]
@stevenschlansker Yeah, I plan on having quite a few RCs out, so by "not next" just means 2-3 week delay. Sort of to isolate things. But I get the point of as-early-as-possible wrt weeding out bugs.

On performance, yeah, +5% is measurable and has some value.

But odd thing is this -- now testing serialization I see no difference whatsoever. That is... odd. Would suggest deserialization speed up would have something to do with setters (parameter) maybe?

Finally: one question on MethodHandles -- I noticed removal of some Field-based classes. I assume field access still exists; how is that now handled?
(I guess I should just spend time actually going through PR :) )

File: src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java [hunk 4]
```diff
@@ -21,19 +33,57 @@ public final class AnnotatedMethod
...
+        boolean staticMethod = Modifier.isStatic(method.getModifiers());
+        int argCount = method.getParameterCount() + (staticMethod ? 0 : 1);
+        _nullaryInvoker = argCount > 0 ? null : invoker.asType(MethodType.methodType(Object.class));
+        _unaryConstructorInvoker = argCount != 1 ? null : invoker.asType(MethodType.methodType(Object.class, Object.class));
+    }
+
+    public static AnnotatedMethod of(TypeResolutionContext ctxt, Method method,
+            AnnotationMap classAnn, AnnotationMap[] paramAnnotations)
+    {
+        if (method == null) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:64]: Can just `Objects.requireNonNull()` since it's internal failure if called with null.

File: src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java [hunk 5]
```diff
@@ -73,27 +123,56 @@ public Class<?> getRawType() {
 
     @Override
     public final Object call() throws Exception {
-        // 31-Mar-2021, tatu: Note! This is faster than calling without arguments
-        //   because JDK in its wisdom would otherwise allocate `new Object[0]` to pass
-        return _method.invoke(null);
+        try {
+            return _nullaryInvoker.invokeExact();
+        } catch (final Throwable e) {
+            throw sneakyThrow(e);
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:129]: I guess MethodHandle.invokeXxx() report throwing `Throwable`, but is there anything beyond `Exception` and `Error` (unchecked)?

**Comment by stevenschlansker** [comment]
> Finally: one question on MethodHandles -- I noticed removal of some Field-based classes. I assume field access still exists; how is that now handled?

MethodHandles unify method calls and field get / set with [unreflectGetter](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#unreflectGetter(java.lang.reflect.Field))
 and [unreflectSetter](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#unreflectSetter(java.lang.reflect.Field)) which present field read / write as trivial getter and setter methods. So we just use that then treat everything as a method now.

File: src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java (hunk 5 above)
Comment (stevenschlansker) [reply src/main/java/tools/jackson/databind/introspect/AnnotatedMethod.java:129]: No, it just is passing through whatever is thrown downstream - there should be no 'new' dynamically exception throwable here, all the access checks were done at MethodHandle time.

**Comment by stevenschlansker** [comment]
> But odd thing is this -- now testing serialization I see no difference whatsoever. That is... odd. Would suggest deserialization speed up would have something to do with setters (parameter) maybe?

It can be really hard to nail down performance on this kind of thing... my understanding from reading the literature is that it should help, but maybe hard to see among all the other work any particular workload does.

It's possible different shapes have different costs. For example when deserializing, `method.invoke(arg)` I think will always allocate an `Object[]` for varargs, whereas `methodHandle.invokeExact(arg)` should not need to do any varargs since it is the magic `@PolymorphicSignature`

File: src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java [hunk 6]
```diff
@@ -935,14 +938,13 @@ protected SettableBeanProperty constructSettableProperty(DeserializationContext
...
         TypeDeserializer typeDeser = (TypeDeserializer) type.getTypeHandler();
         SettableBeanProperty prop;
-        if (mutator instanceof AnnotatedMethod) {
-            prop = new MethodProperty(propDef, type, typeDeser,
-                    beanDesc.getClassAnnotations(), (AnnotatedMethod) mutator);
-        } else {
-            // 08-Sep-2016, tatu: wonder if we should verify it is `AnnotatedField` to be safe?
-            prop = new FieldProperty(propDef, type, typeDeser,
-                    beanDesc.getClassAnnotations(), (AnnotatedField) mutator);
+        if (!ClassUtil.checkAndFixAccess(mutator.getMember(), ctxt.getConfig().isEnabled(MapperFeature.OVERRIDE_PUBLIC_ACCESS_MODIFIERS))) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:941]: I think `ctxt.getConfig().isEnabled(...)` can be shortened to `ctxt.isEnabled(...)` (there should be convenience accessor).

But aside from that, could you add a brief comment explaining reasons why code short-circuits here?
(cannot access I assume but... is that then quietly swallowing cases causing surprises?)

File: src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java [hunk 7]
```diff
@@ -963,6 +965,11 @@ protected SettableBeanProperty constructSettableProperty(DeserializationContext
         return prop;
     }
 
+    private boolean finalField(AnnotatedMember am) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:968]: naming: `finalField()` -> `isFinalField()`

File: src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java [hunk 8]
```diff
@@ -22,23 +24,26 @@ public final class MethodProperty
 {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:24]: Since this now handles Setters and Fields too, class Javadoc needs to be changed?

File: src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java [hunk 9]
```diff
@@ -22,23 +24,26 @@ public final class MethodProperty
...
     private static final long serialVersionUID = 1;
 
-    protected final AnnotatedMethod _annotated;
+    protected final AnnotatedMember _annotated;
 
     /**
      * Setter method for modifying property value; used for
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/impl/MethodProperty.java:30]: This needs changing too if used for Fields as well as Setter methods.

File: src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java [hunk 10]
```diff
@@ -66,24 +66,15 @@ Creators collect(Class<?> primaryMixIn)
...
+                constructor == null || (_collectAnnotations && _intr.hasIgnoreMarker(_config, constructor)));
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java:74]: Is null check needed? It wasn't there earlier?

**Review by cowtowncoder — APPROVED** [review]
Looks pretty good, only minor final touches and I can merge it.

File: src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java (hunk 10 above)
Comment (stevenschlansker) [reply src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java:74]: It's now necessary because access checks have been pushed up from the point of use (when constructor is called) to earlier in deserializer creation (when constructors and factories are collected) so they are now filtered out earlier too. I added a comment.

File: src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java (hunk 6 above)
Comment (stevenschlansker) [reply src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java]: Added a brief comment. I agree, some conditions regarding module visibility could quietly swallow members here, if we cannot call `setAccessible` on them. I think in those cases though the whole operation will fail anyway until you allow Jackson module to open your module with beans in it.

There might be some follow-up tweaking to this logic necessary. I got all the existing tests passing but I am sure there are new exciting corner cases to discover with module visibility.

Come to think of it, can we remove the `MapperFeature.OVERRIDE_PUBLIC_ACCESS_MODIFIERS` now? I believe with MethodHandles used, there should no longer be any performance benefit to calling `setAccessible` unconditionally. I am happy to file a follow-up PR if this is acceptable. If not, we could consider turning it off by default now.

Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java]: Good idea -- I'll file an issue for changing defaults, will merge this PR and we'll go from there.

Actual removal of that feature is an option too, but at least start by changing its default.

**Comment by cowtowncoder** [comment]
@stevenschlansker Ok, looks like this also broke most downstream modules... :-(
Cascading builds at least gives that info in ~15 minutes after merge.

Will file issues, add a notes/refs here.

**Comment by cowtowncoder** [comment]
* Afterburner: https://github.com/FasterXML/jackson-modules-base/issues/294

**Comment by cowtowncoder** [comment]
Ugh. Seems like almost everything downstream fails, more or less catastrophically.

Will likely need to revert this PR later tonight.

**Comment by stevenschlansker** [comment]
:cry:

**Comment by stevenschlansker** [comment]
I'll try to find some time in the next week or two to check out the downstream modules and get those working.
I'm sure it's not trivial to set up, but it could be helpful to have a CI check that all of Jackson builds together.

**Comment by cowtowncoder** [comment]
> I'll try to find some time in the next week or two to check out the downstream modules and get those working. I'm sure it's not trivial to set up, but it could be helpful to have a CI check that all of Jackson builds together.

I have been trying to think of that over time and haven't thought of good maintainable solution.
Cascading rebuild is the first and only improvement -- which at least tightens up feedback loop from days (a week or more, even) to less than an hour.

Resource-wise full rebuild job for all PRs just for databind might be doable, although not sure how heavy resource-wise it'd be wrt Github charging (FasterXML has paid, not free, account).

**Comment by cowtowncoder** [comment]
Will revert now.

**Comment by cowtowncoder** [comment]
Reverted for now. Although failure was wide, it is possible there might be just small number of actual issues so maybe focusing on solving problems by just one downstream repo (like `jackson-dataformat-xml`) would make sense.

File: src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java [hunk 11]
```diff
@@ -935,14 +936,15 @@ protected SettableBeanProperty constructSettableProperty(DeserializationContext
...
         TypeDeserializer typeDeser = (TypeDeserializer) type.getTypeHandler();
         SettableBeanProperty prop;
-        if (mutator instanceof AnnotatedMethod) {
-            prop = new MethodProperty(propDef, type, typeDeser,
-                    beanDesc.getClassAnnotations(), (AnnotatedMethod) mutator);
-        } else {
-            // 08-Sep-2016, tatu: wonder if we should verify it is `AnnotatedField` to be safe?
-            prop = new FieldProperty(propDef, type, typeDeser,
-                    beanDesc.getClassAnnotations(), (AnnotatedField) mutator);
+        // 06-04-2025, scs: we cannot always see members if e.g. they are in a different module that does not
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:939]: Do we need to take a performance hit here? Why not let this fail if access is not setup? Null seems like a hack to me. The Java exceptions are fairly informative about the right add-opens are needed. @cowtowncoder wdyt? This looks like something we could do without. I only have a mobile to check this so the small screen view might not be the right way to check this PR.

Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:939]: Not sure what performance hit you are thinking -- this is not perf sensitive path (it's during deserializer construction).

I agree it seems odd to return `null` but I don't remember reasoning over straight failure.
@stevenschlansker is author here.
One thing to note tho is that `ClassUtil.checkAndFixAccess` just blindly overrides access if I recall and does not attempt to determine if call would likely fail -- it is just based on configuration settings. That could probably be improved, but we cannot test call here (we are not deserializing, we are building deserializer) to make sure.

Comment (stevenschlansker) [reply src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:939]: From memory, the reason for this is because the access checks moved from "Method invoke time" to "MethodHandle create time" - this is one of the potential benefits of MH, lifting access / visibility checks from each-invocation to once-at-creation.

Some JDK classes in the modular world are not visible to user code - you're not allowed to crack open java.lang anymore, for example, without various `--add-opens` directives.

Jackson creates deserializers for types that are not used in some cases (I don't recall which) but basically this is skipping a failure in the case where we see some JDK type that we cannot crack open (Date IIRC?), but it does not matter, because the deserializer does not get used. So we skip an irrelevant failure.

I agree `null` is a bit weird. Improvements welcome, but the code is that way for a reason at least :)

Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/BeanDeserializerFactory.java:939]: Ok. @pjfanning if you wanted to try a PR with explicit fail, we could see if any of unit tests fail? (or just test locally).

@stevenschlansker Yeah I do not doubt there was (is) a reason (I vaguely recall discussions). But also do not remember context well enough to know the "right way" to go.
